### PR TITLE
Add support for converting to CDFG format at load time

### PR DESCRIPTION
**Pull request by ChrisCummins**
This adds a `cdfg` flag to the dataset `load()` operation that converts a ProGraML graph to CDFG format during loading. I've also added a CDFG vocabulary file.

I added a `cdfg` flag to the `POJ104Dataset` constructor to support constructing a CDFG dataset. Care must be taken to ensure that the cached files don't conflict between the CDFG/non-CDFG versions. I think I've done this right but @Zacharias030 I could use a good pair of eyes to check for errors :)

I've included a graph2cdfg binary compiled for Linux, rather than having to build it from scratch.

### CODE REVIEW

File: deeplearning/ml4pl/poj104/dataset.py [hunk 1]
```diff
@@ -789,12 +823,14 @@ def raw_file_names(self):
 
     @property
     def processed_file_names(self):
-        base = f'{self.split}_data.pt'
+        name = f"{self.split}_cdfg" if self.cdfg else self.split
+
+        base = f'{name}_data.pt'
```
Comment (Zacharias030) [inline deeplearning/ml4pl/poj104/dataset.py:828]: Ah great! That's exactly what I wanted to do and where I was worried we'll get undefined behaviour.

File: deeplearning/ml4pl/poj104/dataset.py [hunk 2]
```diff
@@ -30,16 +31,48 @@
...
-        graph.ParseFromString(f.read())
+        proto = f.read()
+
+    if cdfg:
+        graph2cdfg = subprocess.Popen(
+            [str(GRAPH2CDFG), '--stdin_fmt=pb', '--stdout_fmt=pb'],
+            stdin=subprocess.PIPE,  stdout=subprocess.PIPE
+        )
+        proto, _ = graph2cdfg.communicate(proto)
+        assert not graph2cdfg.returncode, f"CDFG conversion failed: {file}"
```
Comment (Zacharias030) [inline deeplearning/ml4pl/poj104/dataset.py:73]: Cool! Thanks a lot for staying in python land with this code base 👍 
It's maybe the second time in my life that I need this 'pipe to subprocess' pattern and nice to see it here so cleanly executed!

File: deeplearning/ml4pl/poj104/dataset.py [hunk 3]
```diff
@@ -848,20 +885,13 @@ def process(self):
...
-        print('\n--- Read data from', data_file)
-        df = pd.read_csv(data_file)
         # Get list of source file names and attributes
-        input_files = df["benchmark"].values   # list of strings of benchmark names
-        dataset = df["dataset"].values         # list of strings of dataset descriptions
-        aux_transfer_size = df["transfer"].values
-        aux_wg_size = df["wgsize"].values
-        oracle = df['oracle'].values
-        runtime_cpu = df['runtime_cpu'].values
-        runtime_gpu = df['runtime_gpu'].values
```
Comment (Zacharias030) [inline deeplearning/ml4pl/poj104/dataset.py:864]: I wonder where the runtimes have gone?
See below

File: deeplearning/ml4pl/poj104/dataset.py [hunk 4]
```diff
@@ -871,42 +901,32 @@ def process(self):
...
+            if self.cdfg:
+                proto = load(filename)
+
+            # Add the features and label.
+            proto_features = proto.features.feature
+            data['y'] = torch.tensor(proto_features["devmap_label"].int64_list.value[0]).view(1)
+            data['aux_in'] = torch.tensor([
+                proto_features["transfer_bytes"].int64_list.value[0],
+                proto_features["wgsize"].int64_list.value[0],
+            ])
```
Comment (Zacharias030) [inline deeplearning/ml4pl/poj104/dataset.py:919]: Do we not get the runtimes for devmap?
I forgot if we even use them, or only accuracy. Did we say aggregated speedups are a meaningless metric?

File: deeplearning/ml4pl/poj104/dataset_test.py [hunk 5]
```diff
@@ -1,14 +1,19 @@
 #!/usr/bin/env python3
 
+import os
 from deeplearning.ml4pl.poj104 import dataset
```
Comment (Zacharias030) [inline deeplearning/ml4pl/poj104/dataset_test.py:4]: I wouldn't mind if you would rather import `pytest` here, but as is looks perfect too.

**Review by Zacharias030 — APPROVED** [review]
Looks great!

File: deeplearning/ml4pl/poj104/dataset.py (hunk 2 above)
Comment (ChrisCummins) [reply deeplearning/ml4pl/poj104/dataset.py:73]: Easier for everyone! The downside is that the binary I committed won't work for everyone. In *theory* it should work on any vaguely-Ubuntu Linux distro, but if you encounter errors you'll have to build it.

File: deeplearning/ml4pl/poj104/dataset.py (hunk 4 above)
Comment (ChrisCummins) [reply deeplearning/ml4pl/poj104/dataset.py:919]: I didn't include runtimes in the exported features, so they're not here :( It would be easy to add them and compute a new dataset, so let me know if you want me to do that. My thinking was that we're only interested in accuracy

File: deeplearning/ml4pl/poj104/dataset_test.py (hunk 5 above)
Comment (ChrisCummins) [reply deeplearning/ml4pl/poj104/dataset_test.py:4]: That would be the "right way" :) at some point I'd like to pull this code back into the main ProGraML repo, will do that then
